src/esp/assets/PhysicsManager.cpp

```diff
@@ -0,0 +1,266 @@
+  _bBroadphase = new btDbvtBroadphase();
+  _bSolver = new btSequentialImpulseConstraintSolver();
+  _bWorld = new btDiscreteDynamicsWorld(_bDispatcher, _bBroadphase, _bSolver,
+                                        _bCollisionConfig);
+
+  // btCollisionDispatcher *dispatcher = static_cast<btCollisionDispatcher
+  // *>(_bWorld.getDispatcher());
+  // btGImpactCollisionAlgorithm::registerAlgorithm(dispatcher);
+
+  _debugDraw.setMode(Magnum::BulletIntegration::DebugDraw::Mode::DrawWireframe);
+  _bWorld->setGravity({0.0f, -10.0f, 0.0f});
+  //_bWorld.setGravity({0.0f, 0.0f, -10.0f});
+  //_bWorld.setDebugDrawer(&_debugDraw);
+
+  physicsNode = node;
+
+  _timeline.start();
+  _initialized = true;
+  _do_profile = do_profile;
+  LOG(INFO) << "Initialized Physics Engine.";
+
+  return true;
+}
+
+PhysicsManager::~PhysicsManager() {
+  LOG(INFO) << "Deconstructing PhysicsManager";
+  if (_initialized) {
+    delete _bCollisionConfig;
+    delete _bDispatcher;
+    delete _bBroadphase;
+    delete _bSolver;
+    delete _bWorld;
+  }
+}
+
+void PhysicsManager::getPhysicsEngine() {}
+
+// Bullet Mesh conversion adapted from:
+// https://github.com/mosra/magnum-integration/issues/20
+bool PhysicsManager::initScene(
+    const AssetInfo& info,
+    const MeshMetaData& metaData,
+    std::vector<CollisionMeshData> meshGroup,
+    physics::BulletRigidObject* physObject) {
+
+  // Test Mesh primitive is valid
+  for (CollisionMeshData& meshData: meshGroup) {
+    if (!isMeshPrimitiveValid(meshData)) {return false;}
+  }
+
+  float mass = 0.0f;    // TODO (JH) weight is currently hardcoded
+  bool objectSuccess;
+  if (info.type == AssetType::INSTANCE_MESH) {
+    // ._semantic.ply mesh data
+    LOG(INFO) << "Initialize instance: before";
+    objectSuccess = physObject->initializeScene(info, mass, meshGroup, *_bWorld);
+    LOG(INFO) << "Initialize instance: after";
+    physObject->syncPose();
+  } 
+  else if (info.type == AssetType::FRL_INSTANCE_MESH) {
+    // FRL mesh
+    LOG(INFO) << "Initialize FRL: before";
+    objectSuccess = physObject->initializeScene(info, mass, meshGroup, *_bWorld);
+    LOG(INFO) << "Initialize FRL: after, success " << objectSuccess;
+    physObject->syncPose();
+  }
+  else {
+    // GLB mesh data
+    LOG(INFO) << "Initialize GLB: before";
+    objectSuccess = physObject->initializeScene(info, mass, meshGroup, *_bWorld);
+    LOG(INFO) << "Initialize GLB: after, success " <<objectSuccess;
+    physObject->syncPose();
+  }
+
+  LOG(INFO) << "Init scene done";
+  return objectSuccess;
+}
+
+
+bool PhysicsManager::initObject(
+    const AssetInfo& info,
+    const MeshMetaData& metaData,
+    std::vector<CollisionMeshData> meshGroup,
+    physics::BulletRigidObject* physObject) {
+
+  // Test Mesh primitive is valid
+  for (CollisionMeshData& meshData: meshGroup) {
+    if (!isMeshPrimitiveValid(meshData)) {return false;}
+  }
+
+  bool objectSuccess;
+  float mass;
```

Comment:
mass = 0.0f.

Are you sure bullet is using float instead of double?
In physically based simulation, it is super rare to use single precision floating point to do the computation. It causes significant numerical instability sometimes.

Reply:
`btScalar` is usually a float, yes (but can be configured to be a double instead (which is very rare OTOH)). This is for the interfaces, not sure what they're using internally for calculations.

Reply:
Bullet author here: I would highly recommend using double precision, that's our default in the context of robotics and reinforcement learning. Instead of implementing your own C++ and Python binding, you may also learn a bit more about [PyBullet](https://docs.google.com/document/d/10sXEhzFRSnvFcl3XxNGhnD4N2SedqwdAvK3dsihxVUA/edit#heading=h.2ye70wns7io3). PyBullet also shows the usage of articulated bodies (btMultiBody) instead of just btRigidBody, and Inverse Kinematics, Inverse Dynamics etc. Happy to discuss more, just catch me on twitter or email.

Reply:
Hey Erwin, thanks for the comment. We are absolutely looking into PyBullet and would love to chat sometime.

Reply:
In PyBullet, it is easy to create more complex robots in btMultiBody (generalized coordinates) or optionally as btRigidBody+btTypeConstraints. Enabling btMultiBody is not easy to get right. Since you already have Python, it may be an easy integration. Happy to chat.

Reply:
You can see some of the features of PyBullet in its quickstart guide. We use it a lot for reinforcement learning and robotics. For example, there is a deep_mimic version, re-implementing Jason Peng's motion imitation, and various quadruped robots and arm grasping setups. See https://docs.google.com/document/d/10sXEhzFRSnvFcl3XxNGhnD4N2SedqwdAvK3dsihxVUA/edit#heading=h.2ye70wns7io3